src/main/host/timer.rs

```diff
@@ -0,0 +1,237 @@
+                on_expire: Box::new(on_expire),
+            })),
+        }
+    }
+
+    pub fn expiration_count(&self) -> u64 {
+        self.magic.debug_check();
+        self.internal.borrow().expiration_count
+    }
+
+    pub fn expire_interval(&self) -> SimulationTime {
+        self.magic.debug_check();
+        self.internal.borrow().expire_interval
+    }
+
+    pub fn consume_expiration_count(&mut self) -> u64 {
+        self.magic.debug_check();
+        let mut internal = self.internal.borrow_mut();
+        let e = internal.expiration_count;
+        internal.expiration_count = 0;
+        e
+    }
+
+    pub fn remaining_time(&self) -> SimulationTime {
+        self.magic.debug_check();
+        let t = if let Some(t) = self.internal.borrow().next_expire_time {
+            t
+        } else {
+            return SimulationTime::ZERO;
+        };
+        let now = Worker::current_time().unwrap();
+        t.saturating_duration_since(&now)
+    }
+
+    pub fn interval(&self) -> SimulationTime {
+        self.magic.debug_check();
+        self.internal.borrow().expire_interval
+    }
+
+    pub fn disarm(&mut self) {
+        self.magic.debug_check();
+        let mut internal = self.internal.borrow_mut();
+        internal.next_expire_time = None;
+        internal.expire_interval = SimulationTime::ZERO;
+        internal.expiration_count = 0;
+        internal.min_valid_expire_id = internal.next_expire_id;
+    }
+
+    fn timer_expire(
+        internal_weak: &Weak<AtomicRefCell<TimerInternal>>,
+        host: &mut Host,
+        expire_id: u64,
+    ) {
+        let internal = if let Some(internal) = Weak::upgrade(internal_weak) {
+            internal
+        } else {
+            trace!("Expired Timer no longer exists.");
+            return;
+        };
+        let mut internal_brw = internal.borrow_mut();
+        trace!(
+            "timer expire check; expireID={} minValidExpireID={}",
+            expire_id,
+            internal_brw.min_valid_expire_id
+        );
+        if expire_id < internal_brw.min_valid_expire_id {
+            // Cancelled.
+            return;
+        }
+
+        let next_expire_time = internal_brw.next_expire_time.unwrap();
+        if next_expire_time > Worker::current_time().unwrap() {
+            // Hasn't expired yet. Check again later.
+            Self::schedule_new_expire_event(&mut *internal_brw, internal_weak.clone(), host);
+            return;
+        }
+
+        internal_brw.expiration_count += 1;
+        if internal_brw.expire_interval > SimulationTime::ZERO {
+            internal_brw.next_expire_time =
+                Some(internal_brw.next_expire_time.unwrap() + internal_brw.expire_interval);
+            Self::schedule_new_expire_event(&mut *internal_brw, internal_weak.clone(), host);
+        }
+
+        // Re-borrow as an immutable reference while executing the callback.
+        drop(internal_brw);
+        let internal_brw = internal.borrow();
```

Comment:
I don't think the `drop` is needed here?

Reply:
This borrow fails without the drop above. IIUC for let bindings that are overwritten/shadowed this way, the "orphaned" object isn't actually dropped until the end of the current scope.